### PR TITLE
Mandd/knapsack models

### PR DESCRIPTION
**Pull request by mandd**
It is needed to develop Knapsack models that can be used coupled with RAVEN when
the desired optimization problem requires the use of specific models to generate
knapsack required parameters.
Ref. Issue #7

### CODE REVIEW

File: src/knapsack/BaseKnapsackModel.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      Method to initialize the BaseKnapsack model
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, runInfoDict, dict, dictionary containing all the RunInfo parameters (XML node <RunInfo>)
+      @ In, inputFiles, list, list of input files (if any)
+      @ Out, None
+    """
+    pass      
+      
+      
+  def run(self, container, Inputs):
```
Comment (wangcj05) [inline src/knapsack/BaseKnapsackModel.py:85]: change "Inputs" to "inputs" or "inputDict"

File: src/knapsack/BaseKnapsackModel.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      @ In, container, object, self-like object where all the variables can be stored
+      @ In, Inputs, dict, dictionary of inputs from RAVEN
+    """   
+    totalValue = 0.0  
+    
+    for key in container.mapping:
+      if key in Inputs.keys() and Inputs[key] in [0.0,1.0]:
+        if Inputs[key] == 1.0:
+          testValue = self.capacity - Inputs[container.mapping[key][1]]
+          if testValue > 0:
```
Comment (wangcj05) [inline src/knapsack/BaseKnapsackModel.py:98]: testValue > 0 --> testValue >= 0?

File: tests/test_BaseKnapsackModel.xml [hunk 3]
```diff
@@ -0,0 +1,110 @@
+<Simulation verbosity="debug">
+  <TestInfo>
+    <name>[]</name>
```
Comment (wangcj05) [inline tests/test_BaseKnapsackModel.xml:3]: update name info

File: src/knapsack/BaseKnapsackModel.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      elif child.tag == 'penaltyFactor':
+        self.penaltyFactor = float(child.text.strip())
+      elif child.tag == 'outcome':
+        self.outcome = child.text.strip()
+      elif child.tag == 'choiceValue':
+        self.choiceValue = child.text.strip()
+      elif child.tag == 'map':
+        container.mapping[child.text.strip()] = [child.get('value'),child.get('cost')]
+      elif child.tag == 'variables':
+        variables = [str(var.strip()) for var in child.text.split(",")]
```
Comment (wangcj05) [inline src/knapsack/BaseKnapsackModel.py:69]: This is the External Model for RAVEN, I think we'd better to use InputData and InputTypes to process the xml input.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
some comments.

File: src/knapsack/BaseKnapsackModel.py (hunk 2 above)
Comment (mandd) [reply src/knapsack/BaseKnapsackModel.py:98]: ouch, good catch!

File: tests/test_BaseKnapsackModel.xml (hunk 3 above)
Comment (mandd) [reply tests/test_BaseKnapsackModel.xml:3]: added

**Review by mandd — COMMENTED** [review]
Addressing reviewer comments

File: src/knapsack/BaseKnapsackModel.py (hunk 4 above)
Comment (wangcj05) [reply src/knapsack/BaseKnapsackModel.py:58]: I have checked other External Models inside LOGOS, they are not using the InputData yet. If you want to use InputData for LOGOS, I suggest we open an issue, and adapt the InputData in another PR.

Comment (mandd) [reply src/knapsack/BaseKnapsackModel.py:58]: Issue #9 has been opened.
We should have a chat on how to fix it.

**Comment by mandd** [comment]
Tests are green on my local machine:
Screen Shot 2021-02-04 at 4 01 50 PM

Documentation compiles as well without errors

**Review by wangcj05 — APPROVED** [review]

**Comment by wangcj05** [comment]
This PR is good to be merged.

**Pull request merged**